libfuzzer/libfuzz_harness.cc

```diff
@@ -95,6 +379,159 @@
+    int program_counter,
+    const uint64_t registers[16],
+    const uint8_t* stack_start,
+    size_t stack_length,
+    uint64_t register_mask,
+    const uint8_t* stack_mask)
+{
+#if defined(UBPF_ENABLE_LIBFUZZER_CONSTRAINT_CHECK)
+    ubpf_context_t* ubpf_context = reinterpret_cast<ubpf_context_t*>(context);
+    UNREFERENCED_PARAMETER(stack_start);
+    UNREFERENCED_PARAMETER(stack_length);
+
+    std::string label = std::to_string(program_counter) + ":-1";
+
+    if (program_counter == 0) {
+        return;
+    }
+
+    // Build set of string constraints from the register values.
+    std::set<std::string> constraints;
+    for (int i = 0; i < 10; i++) {
+        if ((register_mask & (1 << i)) == 0) {
+            continue;
+        }
+        uint64_t reg = registers[i];
+        std::string register_name = "r" + std::to_string(i);
+
+        // Given the register value, classify it as packet, context, stack, or unknown and add the appropriate
+        // constraint.
+        address_type_t type = ubpf_classify_address(ubpf_context, reg);
+        switch (type) {
+        case address_type_t::Packet:
+            constraints.insert(register_name + ".type=packet");
+            constraints.insert(register_name + ".packet_offset=" + std::to_string(reg - ubpf_context->data));
+            constraints.insert(
+                register_name + ".packet_size=" + std::to_string(ubpf_context->data_end - ubpf_context->data));
+            break;
+
+        case address_type_t::Context:
+            constraints.insert(register_name + ".type=ctx");
+            constraints.insert(
+                register_name + ".ctx_offset=" + std::to_string(reg - reinterpret_cast<uint64_t>(ubpf_context)));
+            break;
+
+        case address_type_t::Stack:
+            constraints.insert(register_name + ".type=stack");
+            constraints.insert(register_name + ".stack_offset=" + std::to_string(reg - ubpf_context->stack_start));
+            break;
+
+        case address_type_t::Unknown:
+            constraints.insert("r" + std::to_string(i) + ".uvalue=" + std::to_string(registers[i]));
+            constraints.insert(
+                "r" + std::to_string(i) + ".svalue=" + std::to_string(static_cast<int64_t>(registers[i])));
+            break;
+        }
+    }
+
+    // Call ebpf_check_constraints_at_label with the set of string constraints at this label.
+
+    std::ostringstream os;
+
+    if (!ebpf_check_constraints_at_label(os, label, constraints)) {
+        std::cerr << "Label: " << label << std::endl;
+        std::cerr << os.str() << std::endl;
+        throw std::runtime_error("ebpf_check_constraints_at_label failed");
+    }
+#else
+    UNREFERENCED_PARAMETER(context);
+    UNREFERENCED_PARAMETER(program_counter);
+    UNREFERENCED_PARAMETER(registers);
+    UNREFERENCED_PARAMETER(stack_start);
+    UNREFERENCED_PARAMETER(stack_length);
+    UNREFERENCED_PARAMETER(register_mask);
+    UNREFERENCED_PARAMETER(stack_mask);
+#endif
+}
```

Comment:
This is behind a CMake option as the corresponding PR https://github.com/vbpf/ebpf-verifier/pull/729 that adds certain APIs isn't merge yet. Once it's merged, then it might make sense to make this a runtime option.